File: LinqSampleVB/LinqToObjectSample.vb
''' <summary>
''' LinqToObjectのサンプル(VB.Net)
''' </summary>
Public Class LinqToObjectSample

    ''' <summary>
    ''' フィルタのサンプル
    ''' </summary>
    ''' <param name="src">参照元データ</param>
    ''' <returns>主キーが10以上のデータ</returns>
    Public Shared Function WhereSample(ByVal src As List(Of Commons.LinqToObjectTest)) As List(Of Commons.LinqToObjectTest)

        Return src.Where(Function(item) item.PrimaryKey >= 10).ToList()

    End Function

    ''' <summary>
    ''' ソートのサンプル
    ''' </summary>
    ''' <param name="src">参照元データ</param>
    ''' <returns>主キーの降順でソートされたデータ</returns>
    Public Shared Function OrderSample(ByVal src As List(Of Commons.LinqToObjectTest)) As List(Of Commons.LinqToObjectTest)

        Return src.OrderByDescending(Function(item) item.PrimaryKey).ToList()

    End Function

    ''' <summary>
    ''' ジョイン(内部参照)のサンプル
    ''' </summary>
    ''' <param name="src">参照元データ</param>
    ''' <param name="srcSub">参照元サブデータ</param>
    ''' <returns>参照元データの主キーが参照元サブデータの親主キーが紐づくデータ</returns>
    Public Shared Function JoinSample(ByVal src As List(Of Commons.LinqToObjectTest), ByVal srcSub As List(Of Commons.LinqToObjectTest)) As List(Of Commons.LinqToObjectTest)

        Return src.Join(srcSub, Function(ByVal item As Commons.LinqToObjectTest) item.PrimaryKey,
            Function(ByVal subItem As Commons.LinqToObjectTest) subItem.ParentPrimaryKey,
            Function(item, subItem) New Commons.LinqToObjectTest(
            subItem.PrimaryKey,
            subItem.DispleyOrder,
            item.DataName & ":" & subItem.DataName,
            subItem.ParentPrimaryKey
            )).ToList()

    End Function

    ''' <summary>
    ''' グループ(外部参照)のサンプル
    ''' </summary>
    ''' <param name="src">参照元データ</param>
    ''' <param name="srcSub">参照元サブデータ</param>
    ''' <returns>参照元データの主キーが参照元サブデータの親主キーに存在するか否かを追加したデータ</returns>
    Public Shared Function GroupJoinSample(ByVal src As List(Of Commons.LinqToObjectTest), ByVal srcSub As List(Of Commons.LinqToObjectTest)) As IEnumerable(Of Object)

        Return src.GroupJoin(srcSub, Function(ByVal item As Commons.LinqToObjectTest) item.PrimaryKey,
            Function(ByVal subItem As Commons.LinqToObjectTest) subItem.ParentPrimaryKey,
            Function(item, subItems) New With {
            Key .PrimaryKey = item.PrimaryKey,
            Key .DispleyOrder = item.DispleyOrder,
            Key .DataName = item.DataName,
            Key .existsSubItem = subItems.Any()
            }).ToList()

    End Function

    ''' <summary>
    ''' グルーピングのサンプル
    ''' </summary>
    ''' <param name="src">参照元データ</param>
    ''' <returns>「主キー/10」でグループされたデータ</returns>
    Public Shared Function GroupSample(ByVal src As List(Of Commons.LinqToObjectTest)) As IEnumerable(Of Object)

        Return src.GroupBy(Function(item) Math.Floor(item.PrimaryKey / 10),
            Function(item, element) New With {
            Key .PrimaryKeyPer10 = item.ToString(),
            Key .Count = element.Count()
            }).ToList()

    End Function

End Class

File: LinqSampleCSharp/LinqToObjectSample.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LinqSampleCSharp
{
    /// <summary>
    /// LinqToObjectのサンプル(C#)
    /// </summary>
    public class LinqToObjectSample
    {
        /// <summary>
        /// フィルタのサンプル
        /// </summary>
        /// <param name="src">参照元データ</param>
        /// <returns>主キーが10以上のデータ</returns>
        public static List<Commons.LinqToObjectTest > WhereSample(List<Commons.LinqToObjectTest> src)
        {
            return src.Where((item) => item.PrimaryKey >= 10).ToList();
        }

        /// <summary>
        /// ソートのサンプル
        /// </summary>
        /// <param name="src">参照元データ</param>
        /// <returns>主キーの降順でソートされたデータ</returns>
        public static List<Commons.LinqToObjectTest> OrderSample(List<Commons.LinqToObjectTest> src)
        {
            return src.OrderByDescending((item) => item.PrimaryKey).ToList();
        }

        /// <summary>
        /// ジョイン(内部参照)のサンプル
        /// </summary>
        /// <param name="src">参照元データ</param>
        /// <param name="srcSub">参照元サブデータ</param>
        /// <returns>参照元データの主キーが参照元サブデータの親主キーが紐づくデータ</returns>
        public static List<Commons.LinqToObjectTest> JoinSample(List<Commons.LinqToObjectTest> src, List<Commons.LinqToObjectTest> srcSub)
        {
            return src.Join(srcSub, (srcItem) => srcItem.PrimaryKey, (subSrcItem) => subSrcItem.ParentPrimaryKey,
                            (srcItem, subSrcItem) => 
                                new Commons.LinqToObjectTest(
                                    subSrcItem.PrimaryKey,
                                    subSrcItem.DispleyOrder,
                                    srcItem.DataName + ":" + subSrcItem.DataName,
                                    subSrcItem.ParentPrimaryKey)
                            ).ToList();
        }

        /// <summary>
        /// グループ(外部参照)のサンプル
        /// </summary>
        /// <param name="src">参照元データ</param>
        /// <param name="srcSub">参照元サブデータ</param>
        /// <returns>参照元データの主キーが参照元サブデータの親主キーに存在するか否かを追加したデータ</returns>
        public static IEnumerable<Object> GroupJoinSample(List<Commons.LinqToObjectTest> src, List<Commons.LinqToObjectTest> srcSub)
        {
            return src.GroupJoin(srcSub, (srcItem) => srcItem.PrimaryKey, (subSrcItem) => subSrcItem.ParentPrimaryKey,
                            (srcItem, subSrcItem) =>
                                new {
                                    PrimaryKey = srcItem.PrimaryKey,
                                    DispleyOrder = srcItem.DispleyOrder,
                                    DataName = srcItem.DataName,
                                    existsSubItem = subSrcItem.Any()
                                }
                            ).ToList();
        }

        /// <summary>
        /// グルーピングのサンプル
        /// </summary>
        /// <param name="src">参照元データ</param>
        /// <returns>「主キー/10」でグループされたデータ</returns>
        public static IEnumerable<Object> GroupSample(List<Commons.LinqToObjectTest> src)
        {
            return src.GroupBy((srcItem) => Math.Floor(srcItem.PrimaryKey / 10D),
                            (srcItem, elements) =>
                                new
                                {
                                    PrimaryKeyPer10 = srcItem.ToString(),
                                    Count = elements.Count()
                                }
                            ).ToList();
        }
    }
}
